[libnative-image-agent, OpenJDK] Possibly superfluous usage of GraalAccess to get target arch in DefaultOptionHandler causes NPE

Hello,

This line from commit [01d1fe51d8 GR-33542](https://github.com/graalvm/mandrel/commit/01d1fe51d8b3ee0eb3bd03a79cd7e9fdc43cc4f1#diff-3a0eb3a73c78de90cd1b29be21ad046986e389d5e4be166993659ab548ccd550R210)

```diff
+  Architecture arch = GraalAccess.getOriginalTarget().arch;
```
is causing an NPE during native-agent build when building with stock OpenJDK, both 11.0.12_7 and 11.0.13_5.
There is no such error when LabsJDK is used.

```
[libnative-image-agent:283656]     analysis:  24,205.77 ms,  1.76 GB
2 fatal errors detected:
Fatal error:com.oracle.graal.pointsto.util.AnalysisError$ParsingError: Error encountered while parsing org.graalvm.compiler.replacements.nodes.TypePunnedArrayReadIntrinsic$ReadLong.read(jdk.vm.ci.meta.JavaKind, jdk.vm.ci.meta.JavaKind, java.lang.Object, long) 
Parsing context:
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.typePunnedRead(AMD64TruffleArrayUtilsWithMaskSnippets.java:243)
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.arrayRegionEqualsTypePunnedCmp(AMD64TruffleArrayUtilsWithMaskSnippets.java:183)
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.arrayRegionEqualsTypePunnedLoop(AMD64TruffleArrayUtilsWithMaskSnippets.java:165)
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.arrayRegionEqualsWithMask(AMD64TruffleArrayUtilsWithMaskSnippets.java:145)
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.arrayRegionEqualsWithMaskCCC(AMD64TruffleArrayUtilsWithMaskSnippets.java:218)
   at com.oracle.svm.reflect.AMD64TruffleArrayUtilsWithMaskSnippets_arrayRegionEqualsWithMaskCCC_6ed36d34cb2fced92135a279155d7a76d3b355da_4.invoke(Unknown Source)
   at java.base@11.0.13-ea/java.lang.reflect.Method.invoke(Method.java:566)
   at app//com.oracle.svm.core.containers.Metrics.systemMetrics(Metrics.java:63)
   at app//com.oracle.svm.core.containers.Container.metrics(Container.java:44)

	at com.oracle.graal.pointsto.util.AnalysisError.parsingError(AnalysisError.java:126)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.createTypeFlow(MethodTypeFlow.java:311)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureTypeFlowCreated(MethodTypeFlow.java:282)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:103)
	at com.oracle.graal.pointsto.flow.StaticInvokeTypeFlow.update(InvokeTypeFlow.java:420)
	at com.oracle.graal.pointsto.PointsToAnalysis$2.run(PointsToAnalysis.java:595)
	at com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:188)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:172)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.NullPointerException
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:301)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.createTypeFlow(MethodTypeFlow.java:293)
	... 12 more
Fatal error:com.oracle.graal.pointsto.util.AnalysisError$ParsingError: Error encountered while parsing org.graalvm.compiler.replacements.nodes.TypePunnedArrayReadIntrinsic.read(jdk.vm.ci.meta.JavaKind, jdk.vm.ci.meta.JavaKind, jdk.vm.ci.meta.JavaKind, java.lang.Object, long) 
Parsing context:
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.typePunnedRead(AMD64TruffleArrayUtilsWithMaskSnippets.java:244)
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.arrayRegionEqualsTypePunnedCmp(AMD64TruffleArrayUtilsWithMaskSnippets.java:183)
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.arrayRegionEqualsTypePunnedLoop(AMD64TruffleArrayUtilsWithMaskSnippets.java:165)
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.arrayRegionEqualsWithMask(AMD64TruffleArrayUtilsWithMaskSnippets.java:145)
   at platform/jdk.internal.vm.compiler/org.graalvm.compiler.replacements.amd64.AMD64TruffleArrayUtilsWithMaskSnippets.arrayRegionEqualsWithMaskCCC(AMD64TruffleArrayUtilsWithMaskSnippets.java:218)
   at com.oracle.svm.reflect.AMD64TruffleArrayUtilsWithMaskSnippets_arrayRegionEqualsWithMaskCCC_6ed36d34cb2fced92135a279155d7a76d3b355da_4.invoke(Unknown Source)
   at java.base@11.0.13-ea/java.lang.reflect.Method.invoke(Method.java:566)
   at app//com.oracle.svm.core.containers.Metrics.systemMetrics(Metrics.java:63)
   at app//com.oracle.svm.core.containers.Container.metrics(Container.java:44)

	at com.oracle.graal.pointsto.util.AnalysisError.parsingError(AnalysisError.java:126)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.createTypeFlow(MethodTypeFlow.java:311)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureTypeFlowCreated(MethodTypeFlow.java:282)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:103)
	at com.oracle.graal.pointsto.flow.StaticInvokeTypeFlow.update(InvokeTypeFlow.java:420)
	at com.oracle.graal.pointsto.PointsToAnalysis$2.run(PointsToAnalysis.java:595)
	at com.oracle.graal.pointsto.util.CompletionExecutor.executeCommand(CompletionExecutor.java:188)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$executeService$0(CompletionExecutor.java:172)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.NullPointerException
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:301)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.createTypeFlow(MethodTypeFlow.java:293)
	... 12 more
[libnative-image-agent:283656]      [total]:  38,775.82 ms,  1.76 GB
# Printing build artifacts to: /home/tester/jenkins/workspace/mandrel-master-linux-build/mandrel-java11-21.3.0-deve8ab81148f/lib/svm/macros/native-image-agent-library/../../../libnative-image-agent.build_artifacts.txt
Error: Image build request failed with exit status 1
Exception in thread "main" java.lang.RuntimeException: java.lang.RuntimeException: Failed, exit code: 1
	at OperatingSystem.exec(build.java:1733)
	at build.buildAgents(build.java:188)
	at build.main(build.java:170)
Caused by: java.lang.RuntimeException: Failed, exit code: 1
	at OperatingSystem.exec(build.java:1718)
	at build.buildAgents(build.java:188)
	at build.main(build.java:170)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.execute(Main.java:404)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.run(Main.java:179)
	at jdk.compiler/com.sun.tools.javac.launcher.Main.main(Main.java:119)
```

It is filed under Mandrel's tracker as https://github.com/graalvm/mandrel/issues/287

It seems that the arch could be retrieved without pulling in likely superfluous classes.

(thx @galderz  for the analysis)
`GraalAccess.getOriginalTarget().arch` code calls into the `GraalRuntime` interface, for which exist 2 main implementations: `SubstrateGraalRuntime`and `HotSpotGraalRuntime`. The problem is that, as a result of using that API, at least with OpenJDK 11.0.x (not labs), `HotSpotGraalRuntime` and a load of `org.graalvm.compiler.hotspot` classes end up being part of the points-to analysis and that causes the errors above, e.g.

```
$ grep HotSpotGraalRuntime used_classes_libnative-image-agent_20210906_125301.txt
org.graalvm.compiler.hotspot.HotSpotGraalRuntime
```

I propose a change that doesn't require the use of `GraalRuntime`, hence there's no possibility for the hotspot implementation to make it to the analysis:

```
$ grep hotspot used_classes_libnative-image-agent_20210906_130032.txt
<none>
$
```
i.e.

```patch
--- a/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/DefaultOptionHandler.java
+++ b/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/DefaultOptionHandler.java
@@ -32,11 +32,11 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.regex.Pattern;
 
+import jdk.vm.ci.runtime.JVMCI;
 import org.graalvm.compiler.options.OptionType;
 
 import com.oracle.svm.driver.MacroOption.MacroOptionKind;
 import com.oracle.svm.driver.NativeImage.ArgumentQueue;
-import com.oracle.svm.hosted.c.GraalAccess;
 
 import jdk.vm.ci.aarch64.AArch64;
 import jdk.vm.ci.amd64.AMD64;
@@ -207,7 +207,7 @@ class DefaultOptionHandler extends NativeImage.OptionHandler<NativeImage> {
                 return true;
             case "--list-cpu-features":
                 args.poll();
-                Architecture arch = GraalAccess.getOriginalTarget().arch;
+                Architecture arch = JVMCI.getRuntime().getHostJVMCIBackend().getTarget().arch;
                 if (arch instanceof AMD64) {
                     nativeImage.showMessage("All AMD64 CPUFeatures: " + Arrays.toString(AMD64.CPUFeature.values()));
                     nativeImage.showNewline();
```

--- a/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/DefaultOptionHandler.java
+++ b/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/DefaultOptionHandler.java
@@ -32,11 +32,11 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.regex.Pattern;
 
+import jdk.vm.ci.runtime.JVMCI;
 import org.graalvm.compiler.options.OptionType;
 
 import com.oracle.svm.driver.MacroOption.MacroOptionKind;
 import com.oracle.svm.driver.NativeImage.ArgumentQueue;
-import com.oracle.svm.hosted.c.GraalAccess;
 
 import jdk.vm.ci.aarch64.AArch64;
 import jdk.vm.ci.amd64.AMD64;
@@ -207,7 +207,7 @@ class DefaultOptionHandler extends NativeImage.OptionHandler<NativeImage> {
                 return true;
             case "--list-cpu-features":
                 args.poll();
-                Architecture arch = GraalAccess.getOriginalTarget().arch;
+                Architecture arch = JVMCI.getRuntime().getHostJVMCIBackend().getTarget().arch;
                 if (arch instanceof AMD64) {
                     nativeImage.showMessage("All AMD64 CPUFeatures: " + Arrays.toString(AMD64.CPUFeature.values()));
                     nativeImage.showNewline();

